crates/syn2mas/src/synapse_reader.rs

```diff
@@ -0,0 +1,281 @@
+            });
+        };
+
+        Ok(localpart)
+    }
+}
+
+/// A Synapse boolean.
+/// Synapse stores booleans as 0 or 1, due to compatibility with old SQLite versions
+/// that did not have native boolean support.
+#[derive(Clone, Debug)]
+pub struct SynapseBool(pub bool);
+
+impl<'r> sqlx::Decode<'r, Postgres> for SynapseBool {
+    fn decode(
+        value: <Postgres as sqlx::Database>::ValueRef<'r>,
+    ) -> Result<Self, sqlx::error::BoxDynError> {
+        <i16 as sqlx::Decode<Postgres>>::decode(value)
+            .map(|boolean_int| SynapseBool(boolean_int != 0))
+    }
+}
+
+impl sqlx::Type<Postgres> for SynapseBool {
+    fn type_info() -> <Postgres as sqlx::Database>::TypeInfo {
+        <i16 as sqlx::Type<Postgres>>::type_info()
+    }
+}
+
+/// A timestamp stored as the number of seconds since the Unix epoch.
+/// Note that Synapse stores MOST timestamps as numbers of **milliseconds** since the Unix epoch.
+/// But some timestamps are still stored in seconds.
+#[derive(Clone, Debug)]
+pub struct SecondsTimestamp(pub DateTime<Utc>);
+
+impl<'r> sqlx::Decode<'r, Postgres> for SecondsTimestamp {
+    fn decode(
+        value: <Postgres as sqlx::Database>::ValueRef<'r>,
+    ) -> Result<Self, sqlx::error::BoxDynError> {
+        <i64 as sqlx::Decode<Postgres>>::decode(value).map(|milliseconds_since_epoch| {
+            SecondsTimestamp(DateTime::from_timestamp_nanos(
+                milliseconds_since_epoch * 1_000_000_000,
+            ))
+        })
+    }
+}
+
+impl sqlx::Type<Postgres> for SecondsTimestamp {
+    fn type_info() -> <Postgres as sqlx::Database>::TypeInfo {
+        <i64 as sqlx::Type<Postgres>>::type_info()
+    }
+}
+
+#[derive(Clone, Debug, FromRow)]
+#[enum_def(table_name = "users")]
+pub struct SynapseUser {
+    /// Full User ID of the user
+    pub name: FullUserId,
+    /// Password hash string for the user. Optional (null if no password is set).
+    pub password_hash: Option<String>,
+    /// Whether the user is a Synapse Admin
+    pub admin: SynapseBool,
+    /// Whether the user is deactivated
+    pub deactivated: SynapseBool,
+    /// When the user was created
+    pub creation_ts: SecondsTimestamp,
+    // TODO ...
+    // TODO is_guest
+    // TODO do we care about upgrade_ts (users who upgraded from guest accounts to real accounts)
+}
+
+#[derive(Iden)]
+pub enum ExtraSynapseUserIden {
+    AppserviceId,
+    IsGuest,
+}
+
+/// List of Synapse tables that we should acquire an `EXCLUSIVE` lock on.
+///
+/// This is a safety measure against other processes changing the data underneath our feet.
+/// It's still not a good idea to run Synapse at the same time as the migration.
+// TODO not complete!
+const TABLES_TO_LOCK: &[&str] = &["users"];
+
+/// Number of migratable rows in various Synapse tables.
+/// Used to estimate progress.
+#[derive(Clone, Debug)]
+pub struct SynapseRowCounts {
+    pub users: i64,
+}
+
+pub struct SynapseReader<'c> {
+    conn: &'c mut PgConnection,
```

Comment:
If we used a `Transaction<'c, Postgres>` here, we'd get the `Drop` behaviour of rolling back the transaction

Reply:
This point is still valid as you're already borrowing the connection here, but I'm fine leaving it like that